tests/presets.spec.js

```diff
@@ -740,6 +740,445 @@ describe('Presets', function() {
+              'version = \'1.1.0\'',
+              '',
+              '[dependencies.bar]',
+              'version = \'2.0.0\'',
+              ''
+            ].join('\n'));
+
+            done();
+          });
+        });
+
+      });
+
+      describe('missing Cargo.toml and existing Cargo.lock', function() {
+
+        beforeEach(function() {
+          this.cwd = tmp.dirSync();
+          this.cargoLock = path.join(this.cwd.name, 'Cargo.lock');
+
+          fs.writeFileSync(this.cargoLock, [
+            '[[package]]',
+            'name = "bar"',
+            'version = "2.0.0"',
+            'source = "registry+https://github.com/rust-lang/crates.io-index"',
+            ''
+          ].join('\n'));
+        });
+
+        afterEach(function() {
+          fs.unlinkSync(this.cargoLock);
+          this.cwd.removeCallback();
+        });
+
+        it('should yield an error', function(done) {
+          presets.updateVersion.cargo({}, this.cwd.name, '1.0.1', (error) => {
+            m.chai.expect(error).to.be.an.instanceof(Error);
+            m.chai.expect(error.code).to.equal('ENOENT');
+            done();
+          });
+        });
+
+      });
+
+      describe('updating Cargo.toml with unchanged version', function() {
+
+        beforeEach(function() {
+          this.cwd = tmp.dirSync();
+          this.cargoToml = path.join(this.cwd.name, 'Cargo.toml');
+
+          fs.writeFileSync(this.cargoToml, [
+            '[package]',
+            'name = "foo"',
+            'version = "1.2.3"',
+            '',
+            '[dependencies.bar]',
+            'version = "2.0.0"',
+            ''
+          ].join('\n'));
+        });
+
+        afterEach(function() {
+          fs.unlinkSync(this.cargoToml);
+          this.cwd.removeCallback();
+        });
+
+        it('should be able to preserve the version', function(done) {
+          presets.updateVersion.cargo({}, this.cwd.name, '1.2.3', (error) => {
+            m.chai.expect(error).to.not.exist;
+
+            m.chai.expect(fs.readFileSync(this.cargoToml, 'utf8')).to.equal([
+              '[package]',
+              'name = "foo"',
+              'version = "1.2.3"',
+              '',
+              '[dependencies.bar]',
+              'version = "2.0.0"',
+              ''
+            ].join('\n'));
+
+            done();
+          });
+        });
+
+      });
+
+      describe('Cargo.toml with a non-semver version', function() {
+
+        beforeEach(function() {
+          this.cwd = tmp.dirSync();
+          this.cargoToml = path.join(this.cwd.name, 'Cargo.toml');
+
+          fs.writeFileSync(this.cargoToml, [
+            '[package]',
+            'name = "foo"',
+            'version = "not-a-semver"',
+            ''
+          ].join('\n'));
+        });
+
+        afterEach(function() {
+          fs.unlinkSync(this.cargoToml);
+          this.cwd.removeCallback();
+        });
+
+        it('should be able to correct the version', function(done) {
```

Comment:
minor nit, but my gut-feeling is that this should say "update the version" rather than "correct the version"

Reply:
...and to make this test more comprehensive, should it be updating both Cargo.lock and Cargo.toml that had `version = "not-a-semver"` ?

Reply:
...actually, I guess there's nothing that restricts the version we're _setting_ to be a semver either, so perhaps this test (or a new one) should be `it('should work with non-semver versions', ` and then update the version from `not-a-semver` to `also-not-a-semver` ? :smiley:

Reply:
Let's try that!

Reply:
It does not work with non-sem-versions. We have a `semver.clean(version)` call at the beginning of the function. This means however that we need to have a test for this.

Reply:
> It does not work with non-sem-versions. We have a semver.clean(version) call at the beginning of the function.

I guess _strictly_ speaking we should only do that if `https://github.com/resin-io/versionist#incrementversion-functionstring` is set to `semver`. But chances are that `versionist` will never get used with non-semver versioning, so we _probably_ don't need to worry about that.

OTOH, do we actually need `semver.clean`? Will versionist ever generate a 'non-clean' semver version?

Reply:
That's interesting. I will trace through the code what the situation is tomorrow and get back to you on this. It looks like `incrementVersion` is not respected. I initially copied this from the `npm` preset, so it should have the same issue likewise (+ the new python handling).

Reply:
@lurch @hedss I finished looking into the code:

We do not support non-semver based schemes. If we need to support that we have to add more hooks: e.g. one for reading current version, etc.

`incrementVersion` is for augmenting semver versions in a custom way. When the feature was added the description reads: `This hook can be declared to customise how a version is incremented given an increment level. It can be handy in some edge cases, and defaults to incrementing a version based on semver rules.`.

There are not any tests/usages of `incrementVersion` throughout the code which use a custom function, so it looks like it was just a nice-to-have idea. My recommendation is for removing it as a customization option, unless we have projects that are practically using it. I am not sure that should be the case though, as semver itself provides a lot of extra flexibility, e.g. `semver.inc('1.2.0', 'prepatch') == '1.2.1-0'` or `semver.inc('1.2.3-alpha.0', 'prerelease') == '1.2.3-alpha.1'`.

Our update presets receive already cleaned versions after the incrementation step. This is how `semver.inc` works itself: `semver.inc('v5.3.4', 'patch') == '5.3.5'`. My thinking is that in the update functions we should accept only clean versions and treat any unclean version as an error.

Thinking about that further: a clean version has to be produced at the very first step while reading current version from the repo and all consecutive operations should expect a clean version and should treat unclean input as error (in terms of being more defensive towards the input correctness).

Given all of this, I think those problems should be treated in separate issues and we can leave these aspect as-is in this particular commit. Looking forward to hearing your opinions and ideas.

Reply:
I'm not sure... The README says "Versionist is non-opinionated. It adapts to your commit practices" so I dunno if we want to explicitly restrict it to only operating on semvers, or if we want to "leave the door open" for other versioning schemes in future? I think the original intention of versionist was for it to be a totally generic tool that any project could use, but I don't know if it's ever been used outside of Resin?
@jviotti As the original author, what do you think about restricting versionist to be semver-only and getting rid of the `incrementVersion` option?
Maybe this is something that should be discussed in a process call @lekkas ?
(For the record, I have no strong opinions either way. Although IMHO it makes sense for the code in the update presets to be as simple as possible, so if there's anything that's being done in _every_ update preset, then perhaps it makes sense to move that code outside the update presets.)